File: packages/python/elizaos/features/advanced_capabilities/providers/facts.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State


async def get_facts_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    sections: list[str] = []
    facts_list: list[dict[str, str]] = []

    entity_id = message.entity_id
    room_id = message.room_id

    if entity_id:
        sender_facts = await runtime.get_memories(
            entity_id=entity_id,
            memory_type="fact",
            limit=10,
        )

        if sender_facts:
            sender = await runtime.get_entity(entity_id)
            sender_name = sender.name if sender and sender.name else "User"
            sections.append(f"\n## Facts about {sender_name}")

            for fact in sender_facts:
                if fact.content and fact.content.text:
                    fact_text = fact.content.text
                    if len(fact_text) > 200:
                        fact_text = fact_text[:200] + "..."
                    facts_list.append(
                        {"entityId": str(entity_id), "entityName": sender_name, "fact": fact_text}
                    )
                    sections.append(f"- {fact_text}")

    if room_id:
        room_facts = await runtime.get_memories(
            room_id=room_id,
            memory_type="fact",
            limit=5,
        )

        if room_facts:
            sections.append("\n## Room Context Facts")
            for fact in room_facts:
                if fact.content and fact.content.text:
                    fact_text = fact.content.text
                    if len(fact_text) > 200:
                        fact_text = fact_text[:200] + "..."
                    facts_list.append({"roomId": str(room_id), "fact": fact_text})
                    sections.append(f"- {fact_text}")

    context_text = ""
    if sections:
        context_text = "# Known Facts" + "\n".join(sections)

    return ProviderResult(
        text=context_text,
        values={
            "factCount": len(facts_list),
            "hasFacts": len(facts_list) > 0,
        },
        data={
            "facts": facts_list,
        },
    )


facts_provider = Provider(
    name="FACTS",
    description="Provides known facts about entities learned through conversation",
    get=get_facts_context,
    dynamic=True,
)

File: packages/rust/src/features/basic_capabilities/bootstrap/providers/facts.rs
//! FACTS provider implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, MemoryType, ProviderResult, State};

use super::Provider;

static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("FACTS"));

/// Provider for known facts about entities.
pub struct FactsProvider;

#[async_trait]
impl Provider for FactsProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(true)
    }

    async fn get(
        &self,
        runtime: &dyn IAgentRuntime,
        message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let mut sections = Vec::new();
        let mut facts_list: Vec<serde_json::Value> = Vec::new();

        let entity_id = message.entity_id;
        let room_id = message.room_id;

        // Get facts about the sender
        if let Some(entity_id) = entity_id {
            match runtime
                .get_memories(None, Some(entity_id), Some(MemoryType::Fact), 10)
                .await
            {
                Ok(facts) => {
                    if !facts.is_empty() {
                        let sender = runtime.get_entity(entity_id).await.ok().flatten();
                        let sender_name = sender
                            .and_then(|s| s.name)
                            .unwrap_or_else(|| "User".to_string());

                        sections.push(format!("\n## Facts about {}", sender_name));

                        for fact in facts {
                            if !fact.content.text.is_empty() {
                                let mut fact_text = fact.content.text.clone();
                                if fact_text.len() > 200 {
                                    fact_text.truncate(200);
                                    fact_text.push_str("...");
                                }

                                facts_list.push(serde_json::json!({
                                    "entityId": entity_id.to_string(),
                                    "entityName": sender_name,
                                    "fact": fact_text
                                }));

                                sections.push(format!("- {}", fact_text));
                            }
                        }
                    }
                }
                Err(e) => {
                    runtime.log_warning("provider:facts", &format!("Error fetching facts: {}", e));
                }
            }
        }

        // Get facts about the room context
        if let Some(room_id) = room_id {
            match runtime
                .get_memories(Some(room_id), None, Some(MemoryType::Fact), 5)
                .await
            {
                Ok(facts) => {
                    if !facts.is_empty() {
                        sections.push("\n## Room Context Facts".to_string());

                        for fact in facts {
                            if !fact.content.text.is_empty() {
                                let mut fact_text = fact.content.text.clone();
                                if fact_text.len() > 200 {
                                    fact_text.truncate(200);
                                    fact_text.push_str("...");
                                }

                                facts_list.push(serde_json::json!({
                                    "roomId": room_id.to_string(),
                                    "fact": fact_text
                                }));

                                sections.push(format!("- {}", fact_text));
                            }
                        }
                    }
                }
                Err(e) => {
                    runtime.log_warning(
                        "provider:facts",
                        &format!("Error fetching room facts: {}", e),
                    );
                }
            }
        }

        let context_text = if sections.is_empty() {
            String::new()
        } else {
            format!("# Known Facts{}", sections.join("\n"))
        };

        Ok(ProviderResult::new(context_text)
            .with_value("factCount", facts_list.len() as i64)
            .with_value("hasFacts", !facts_list.is_empty())
            .with_data("facts", facts_list))
    }
}
